### PR TITLE
Add outcome of bindingoverrides

### PR DESCRIPTION
**Pull request by yiraeChristineKim**
As a policy user that uses `bindingOverrides.remediationAction` on a `PlacementBinding`, I'd like to see in the console the correct remediationAction reflected.

- The policy Results tab's table should have a `Remediation` column. Add an indicator that it is overridden.

- The "Policies" tab's table's `Remediation` column should indicate that there are some overrides. For example, if the parent policy has `remediationAction` set to `inform` but there are some overrides, the text should go from `inform` to `inform/enforce (overridden)`.

Ref: https://issues.redhat.com/browse/ACM-6634

### CODE REVIEW

File: frontend/src/routes/Governance/components/PolicyActionDropdown.tsx [hunk 1]
```diff
@@ -150,10 +156,12 @@ export function PolicyActionDropdown(props: {
       {
         id: 'inform-policy',
         text: t('policy.table.actions.inform'),
-        tooltip: policyRemediationAction === 'inform' ? t('Already informing') : t('Inform policy'),
+        tooltip: policyRemediationAction?.includes(REMEDIATION_ACTION.INFORM)
```
Comment (yiraeChristineKim) [inline frontend/src/routes/Governance/components/PolicyActionDropdown.tsx:159]: This part I am not 100% sure this so I need to double-check with reviewers

File: frontend/src/routes/Governance/components/PolicyActionDropdown.tsx [hunk 2]
```diff
@@ -150,10 +156,12 @@ export function PolicyActionDropdown(props: {
...
+          ? t('Already informing')
+          : t('Inform policy'),
         addSeparator: true,
-        isAriaDisabled: policyRemediationAction === 'inform',
-        click: (item: PolicyTableItem) => {
+        isAriaDisabled: policyRemediationAction?.includes(REMEDIATION_ACTION.INFORM),
```
Comment (yiraeChristineKim) [inline frontend/src/routes/Governance/components/PolicyActionDropdown.tsx:163]: This too

File: frontend/src/routes/Governance/governance.sharedMocks.tsx [hunk 3]
```diff
@@ -224,6 +224,105 @@ const policy1: Policy = {
   },
 }
 
+export const policy2: Policy = {
```
Comment (yiraeChristineKim) [inline frontend/src/routes/Governance/governance.sharedMocks.tsx:227]: These policies are just for unit-test

File: frontend/src/resources/policy.ts [hunk 4]
```diff
@@ -15,6 +15,14 @@ export const PolicyDefinition: IResourceDefinition = {
   kind: PolicyKind,
 }
 
+export enum REMEDIATION_ACTION {
+  ENFORCE_OVERRIDDEN = 'enforce (overridden)',
+  INFORM_ENFORCE_OVERRIDEN = 'inform/enforce (overridden)',
```
Comment (mprahl) [inline frontend/src/resources/policy.ts:20]:
```suggestion
  INFORM_ENFORCE_OVERRIDDEN = 'inform/enforce (overridden)',
```

File: frontend/src/resources/policy.ts [hunk 5]
```diff
@@ -15,6 +15,14 @@ export const PolicyDefinition: IResourceDefinition = {
...
+  INFORM = 'inform',
+  ENFORCE = 'enforce',
+  INFORM_ENFORCE = 'inform/enforce',
+  OVERRIDEN = '(overridden)',
```
Comment (mprahl) [inline frontend/src/resources/policy.ts:24]:
```suggestion
  OVERRIDDEN = '(overridden)',
```

File: frontend/src/resources/policy.ts [hunk 6]
```diff
@@ -30,6 +38,7 @@ export interface Policy {
     placement?: { placementBinding: string; placementRule?: string; placement?: string; policySet?: string }[]
     status?: { clustername: string; clusternamespace: string; compliant?: string }[]
   }
+  remediationResult?: REMEDIATION_ACTION | string
```
Comment (mprahl) [inline frontend/src/resources/policy.ts:41]: Could you please add a comment here explaining what this means?

File: frontend/src/routes/Governance/common/util.tsx [hunk 7]
```diff
@@ -463,30 +464,74 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
   }
-  if (policy.spec?.remediationAction) {
-    return policy.spec.remediationAction
-  }
   const templates = policy.spec['policy-templates']
+  const rootRA = policy.spec.remediationAction
+  if (rootRA === REMEDIATION_ACTION.ENFORCE) {
+    return rootRA
+  }
+  // NOTE RA means Remediation Action
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:476]: It'd be more helpful if this was above `const rootRA....`

File: frontend/src/resources/placement-binding.ts [hunk 8]
```diff
@@ -13,6 +13,10 @@ export interface PlacementBinding extends IResource {
   apiVersion: PlacementBindingApiVersionType
   kind: PlacementBindingKindType
   metadata: Metadata
+  bindingOverrides?: {
+    remediationAction?: 'inform' | 'enforce' | null
```
Comment (mprahl) [inline frontend/src/resources/placement-binding.ts:17]: You can't set an override of "inform". We do accept `Enforce` and `enforce` though.
```suggestion
    remediationAction?: 'Enforce' | 'enforce' | null
```

File: frontend/src/routes/Governance/common/util.tsx [hunk 9]
```diff
@@ -463,30 +464,74 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
-    const templateRemediation = template.objectDefinition.spec.remediationAction
-    if (remediationAggregation === 'inform/enforce') {
-      return
-    } else if (remediationAggregation !== '-' && remediationAggregation !== templateRemediation) {
-      remediationAggregation = 'inform/enforce'
-      return
-    } else if (remediationAggregation !== templateRemediation && templateRemediation) {
-      remediationAggregation = templateRemediation
-      return
+  propagatedPolicies.forEach((propaPolicy) => {
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:478]: This code section looks right but I'll look at it again tomorrow.

File: frontend/src/routes/Governance/common/util.tsx [hunk 10]
```diff
@@ -463,30 +464,74 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
+    const propaPolicyRA = propaPolicy.spec.remediationAction
+    if (propaPolicyRA !== rootRA) {
+      if (remediationAggregation === '-') {
+        remediationAggregation = REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
+      } else if (!remediationAggregation.includes(REMEDIATION_ACTION.ENFORCE_OVERRIDDEN)) {
+        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:484]: If you used a `for` loop instead of `.forEach` you could break here to avoid additional iterations for nothing.

File: frontend/src/routes/Governance/common/util.tsx [hunk 11]
```diff
@@ -463,30 +464,74 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
+      }
+    } else if (propaPolicyRA != undefined) {
+      if (remediationAggregation === '-') {
+        remediationAggregation = REMEDIATION_ACTION.INFORM
+      } else if (remediationAggregation == REMEDIATION_ACTION.ENFORCE_OVERRIDDEN) {
+        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:490]: If you used a `for` loop instead of `.forEach` you could break here to avoid additional iterations for nothing.

Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:490]: what you mean additional iterations?

File: frontend/src/routes/Governance/common/util.tsx [hunk 12]
```diff
@@ -463,30 +464,74 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
+      }
     }
   })
+
+  let remediationAggregationTemp = '-'
+
+  if (!rootRA) {
```
[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]: I tried to avoid additional procedure when root remediation is not null, we don't need to calculate policy-template when we have  root remediation

File: frontend/src/routes/Governance/common/util.tsx (hunk 11 above)
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:490]: I mean that `.forEach` doesn't allow you to stop the loop early. If you use a for loop, you can use `break` after the `remediationAggregation` is set to `inform/enforce (overidden)` since no other value can be set.

Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:490]: For efficiency since after `remediationAggregation` is set to `inform/enforce (overidden)` no other value can possibly be set.

Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:490]: Oh I see!! ok I will change to `for`

File: frontend/src/routes/Governance/common/util.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:497]: @yiraeChristineKim I think you misunderstood my comment. Here is what I meant:

```typescript
  const rootRA = policy.spec.remediationAction
  if (rootRA === REMEDIATION_ACTION.ENFORCE) {
    return rootRA
  } else if (!rootRA) {
    let remediationAggregationTemp = '-'

    templates?.forEach((template: PolicyTemplate) => {
      const templateRemediation = template.objectDefinition.spec.remediationAction
      if (remediationAggregationTemp === REMEDIATION_ACTION.INFORM_ENFORCE) {
        return
      } else if (remediationAggregationTemp !== '-' && remediationAggregationTemp !== templateRemediation) {
        remediationAggregationTemp = REMEDIATION_ACTION.INFORM_ENFORCE
      } else if (remediationAggregationTemp !== templateRemediation && templateRemediation) {
        remediationAggregationTemp = templateRemediation
      }
    })

    if (remediationAggregationTemp === '-') {
      return remediationAggregationTemp
    }

    rootRA = remediationAggregationTemp
}


propagatedPolicies.forEach((propaPolicy) => {
....
```

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]: What about this case, root:null A: enforce B: enforce, template[0]: enforce template[1]: enforce, we expect this case `enforce`, but this logic result is enforce override

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]: Oh also root:null A: enforce B: null, template[0]: inform template[1]: inform,  we expect `inforce/enforce overridden `but actual result is `enforce overridden`

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]: Case root: null, clusterA: enforce, clusterB: null, tmp[0] enforce  tmp[1] enforce, we expect: inform/enfoce (overridden), but actual is enforce (overridden)

we need to change   propagatedPolicies.forEach((propaPolicy) => {

[the remark this answers is not in the document]
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:497]: That example above should be `enforce` because the root policy has no remediation action and both templates are `enforce`, so it's the equivalent of the root policy having `remediationAction: enforce`.

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]: I just unit test with the logic the result is ` inform/enforce (overridden)`

[the remark this answers is not in the document]
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:497]: @yiraeChristineKim are you able to debug a bit to see why?

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]:
```
  propagatedPolicies.forEach((propaPolicy) => {
    const propaPolicyRA = propaPolicy.spec.remediationAction
    if (propaPolicyRA !== rootRA) {
      if (remediationAggregation === '-') {
        remediationAggregation = REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
      } else if (!remediationAggregation.includes(REMEDIATION_ACTION.ENFORCE_OVERRIDDEN)) {
        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
      }
    } else if (propaPolicyRA != undefined) {
      if (remediationAggregation === '-') {
        remediationAggregation = REMEDIATION_ACTION.INFORM
      } else if (remediationAggregation == REMEDIATION_ACTION.ENFORCE_OVERRIDDEN) {
        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
      }
    }
```
enforce == enforce same so it goes to `else if `.  remediationAggregation = inform. The ext turn is undefined and enforce so these two don't match, it falls to  
```
else if (!remediationAggregation.includes(REMEDIATION_ACTION.ENFORCE_OVERRIDDEN)) {
        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
      }
```

[the remark this answers is not in the document]
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:497]: Another tests also failed

[the remark this answers is not in the document]
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:497]: This is what I propose:
```diff
diff --git a/frontend/src/routes/Governance/common/util.test.tsx b/frontend/src/routes/Governance/common/util.test.tsx
index 66e0f7fc..4ef8a654 100644
--- a/frontend/src/routes/Governance/common/util.test.tsx
+++ b/frontend/src/routes/Governance/common/util.test.tsx
@@ -256,7 +256,7 @@ describe('Test getPolicyRemediation', () => {
       ]
       rootPolicy.spec['policy-templates'] = rootPolicyTemplates
       expect(getPolicyRemediation(rootPolicy, [clusterA, clusterB, clusterC])).toEqual(
-        REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
+        REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
       )
     })
     test('getPolicyRemediation should be inform/enforce when template-remediations mixed', () => {
@@ -284,7 +284,7 @@ describe('Test getPolicyRemediation', () => {
       ]
       rootPolicy.spec['policy-templates'] = rootPolicyTemplates
       expect(getPolicyRemediation(rootPolicy, [clusterA, clusterB, clusterC])).toEqual(
-        REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
+        REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
       )
     })
     test('getPolicyRemediation should be enforce when all template-remediations = enforce', () => {
diff --git a/frontend/src/routes/Governance/common/util.tsx b/frontend/src/routes/Governance/common/util.tsx
index 8980ed64..048bdf89 100644
--- a/frontend/src/routes/Governance/common/util.tsx
+++ b/frontend/src/routes/Governance/common/util.tsx
@@ -468,59 +468,64 @@ export function getPolicyRemediation(policy: Policy | undefined, propagatedPolic
   if (!policy) {
     return ''
   }
+
   const templates = policy.spec['policy-templates']
-  const rootRA = policy.spec.remediationAction
+  let rootRA = policy.spec.remediationAction
   if (rootRA === REMEDIATION_ACTION.ENFORCE) {
     return rootRA
   }
-  // NOTE RA means Remediation Action
-  let remediationAggregation = '-'
-  propagatedPolicies.forEach((propaPolicy) => {
-    const propaPolicyRA = propaPolicy.spec.remediationAction
-    if (propaPolicyRA !== rootRA) {
-      if (remediationAggregation === '-') {
-        remediationAggregation = REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
-      } else if (!remediationAggregation.includes(REMEDIATION_ACTION.ENFORCE_OVERRIDDEN)) {
-        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
-      }
-    } else if (propaPolicyRA != undefined) {
-      if (remediationAggregation === '-') {
-        remediationAggregation = REMEDIATION_ACTION.INFORM
-      } else if (remediationAggregation == REMEDIATION_ACTION.ENFORCE_OVERRIDDEN) {
-        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
-      }
-    }
-  })
-
-  let remediationAggregationTemp = '-'
 
   if (!rootRA) {
     templates?.forEach((template: PolicyTemplate) => {
       const templateRemediation = template.objectDefinition.spec.remediationAction
-      if (remediationAggregationTemp === REMEDIATION_ACTION.INFORM_ENFORCE) {
+
+      if (!templateRemediation) {
         return
-      } else if (remediationAggregationTemp !== '-' && remediationAggregationTemp !== templateRemediation) {
-        remediationAggregationTemp = REMEDIATION_ACTION.INFORM_ENFORCE
-      } else if (remediationAggregationTemp !== templateRemediation && templateRemediation) {
-        remediationAggregationTemp = templateRemediation
+      }
+
+      if (rootRA === REMEDIATION_ACTION.INFORM_ENFORCE) {
+        return
+      }
+
+      if (!rootRA) {
+        rootRA = templateRemediation
+      } else if (rootRA !== templateRemediation) {
+        rootRA = REMEDIATION_ACTION.INFORM_ENFORCE
       }
     })
+  }
+
+  let remediationOverride = ''
+  let allOverridden = propagatedPolicies.length !== 0
+  propagatedPolicies.forEach((propaPolicy) => {
+    const propaPolicyRA = propaPolicy.spec.remediationAction
+    if (!propaPolicyRA) {
+      allOverridden = false
 
-    const notOverriden = propagatedPolicies.every((p) => p.spec.remediationAction == undefined)
-    if (notOverriden) {
-      remediationAggregation = remediationAggregationTemp
+      return
     }
 
-    if (remediationAggregation === REMEDIATION_ACTION.ENFORCE_OVERRIDDEN) {
-      if (remediationAggregationTemp === REMEDIATION_ACTION.ENFORCE) {
-        remediationAggregation = REMEDIATION_ACTION.ENFORCE
-      } else {
-        remediationAggregation = REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
-      }
+    if (propaPolicyRA === rootRA) {
+      allOverridden = false
+
+      return
     }
+
+    if (!remediationOverride) {
+      remediationOverride = REMEDIATION_ACTION.ENFORCE
+    }
+  })
+
+  if (allOverridden) {
+    return REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
+  }
+
+  // No overrides are present
+  if (!remediationOverride) {
+    return rootRA
   }
 
-  return remediationAggregation
+  return REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
 }
 
 export function getPropagatedPolicies(rootPolicy: Policy, propaPolicies: Policy[]) {
```


For an easier view, this is the function code I propose:

```typescript
export function getPolicyRemediation(policy: Policy | undefined, propagatedPolicies: Policy[]) {
  if (!policy) {
    return ''
  }

  const templates = policy.spec['policy-templates']
  let rootRA = policy.spec.remediationAction
  if (rootRA === REMEDIATION_ACTION.ENFORCE) {
    return rootRA
  }

  if (!rootRA) {
    templates?.forEach((template: PolicyTemplate) => {
      const templateRemediation = template.objectDefinition.spec.remediationAction

      if (!templateRemediation) {
        return
      }

      if (rootRA === REMEDIATION_ACTION.INFORM_ENFORCE) {
        return
      }

      if (!rootRA) {
        rootRA = templateRemediation
      } else if (rootRA !== templateRemediation) {
        rootRA = REMEDIATION_ACTION.INFORM_ENFORCE
      }
    })
  }

  let remediationOverride = ''
  let allOverridden = propagatedPolicies.length !== 0
  propagatedPolicies.forEach((propaPolicy) => {
    const propaPolicyRA = propaPolicy.spec.remediationAction
    if (!propaPolicyRA) {
      allOverridden = false

      return
    }

    if (propaPolicyRA === rootRA) {
      allOverridden = false

      return
    }

    if (!remediationOverride) {
      remediationOverride = REMEDIATION_ACTION.ENFORCE
    }
  })

  if (allOverridden) {
    return REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
  }

  // No overrides are present
  if (!remediationOverride) {
    return rootRA
  }

  return REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDEN
}
```

File: frontend/src/routes/Governance/policies/Policies.tsx [hunk 13]
```diff
@@ -708,7 +706,7 @@ export default function PoliciesPage() {
           { label: t('Inform/Enforce'), value: 'inform/enforce' },
         ],
         tableFilterFn: (selectedValues, item) => {
-          const policyRemediation = getPolicyRemediation(item.policy)
+          const policyRemediation = item.policy.remediationResult?.replace(' (overridden)', '') || ''
```
Comment (yiraeChristineKim) [inline frontend/src/routes/Governance/policies/Policies.tsx:709]: @mprahl This change is for filter

File: frontend/src/routes/Governance/common/util.tsx [hunk 14]
```diff
@@ -463,28 +463,72 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
+
+  // No overrides are present
+  if (!remediationOverride) {
+    return rootRA
+  }
+
+  return REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDDEN
+}
+
+export function getPropagatedPolicies(rootPolicy: Policy, propaPolicies: Policy[]) {
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:525]: I don't this function being called. Could you please remove it?

File: frontend/src/routes/Governance/components/PolicyActionDropdown.tsx (hunk 1 above)
Comment (mprahl) [reply frontend/src/routes/Governance/components/PolicyActionDropdown.tsx:159]: I think the existing code was fine for this. Could you please change it back?

File: frontend/src/routes/Governance/components/PolicyActionDropdown.tsx [hunk 15]
```diff
@@ -184,15 +192,21 @@ export function PolicyActionDropdown(props: {
...
-          policyRemediationAction === 'inform'
+          policyRemediationAction === REMEDIATION_ACTION.INFORM
             ? undefined
             : [rbacPatch(PolicyDefinition, item.policy.metadata.namespace)],
       },
       {
         id: 'enforce-policy',
         text: t('policy.table.actions.enforce'),
-        tooltip: policyRemediationAction === 'enforce' ? t('Already enforcing') : t('Enforce policy'),
-        isAriaDisabled: policyRemediationAction === 'enforce',
```
Comment (mprahl) [inline frontend/src/routes/Governance/components/PolicyActionDropdown.tsx:195]: Similar comment as before. The existing code was fine. The override shouldn't influence your ability to set the remediationAction on the policy.

File: frontend/src/routes/Governance/policies/Policies.test.tsx [hunk 16]
```diff
@@ -120,6 +126,32 @@ describe('Policies Page', () => {
       )
     )
   })
+
+  test('should show enforce fitler without (overriden)', async () => {
```
Comment (mprahl) [inline frontend/src/routes/Governance/policies/Policies.test.tsx:130]:
```suggestion
  test('should show enforce filter without (overridden)', async () => {
```

File: frontend/src/routes/Governance/policies/policy-details/PolicyDetailsOverview.test.tsx [hunk 17]
```diff
@@ -50,6 +60,35 @@ describe('Policy Details Results', () => {
...
+          }}
+        >
+          <MemoryRouter>
+            <PolicyDetailsOverview policy={mockPolicy[0]} />
+          </MemoryRouter>
+        </RecoilRoot>
+      )
+
+      // wait page load
+      await waitForText(expected)
```
Comment (mprahl) [inline frontend/src/routes/Governance/policies/policy-details/PolicyDetailsOverview.test.tsx:88]: Could you please make this selector more specific? This just checks for any text on the screen.

File: package-lock.json [hunk 18]
```diff
@@ -1413,9 +1413,9 @@
       }
     },
     "node_modules/yaml": {
-      "version": "2.1.3",
-      "resolved": "https://registry.npmjs.org/yaml/-/yaml-2.1.3.tgz",
-      "integrity": "sha512-AacA8nRULjKMX2DvWvOAdBZMOfQlypSFkjcOcu9FalllIDJ1kvlREzcdIZmidQUqqeMv7jorHjq2HlLv/+c2lg==",
+      "version": "2.3.2",
```
Comment (mprahl) [inline package-lock.json:1416]: Are these changes necessary?

File: frontend/src/routes/Governance/policies/Policies.test.tsx [hunk 19]
```diff
@@ -120,6 +126,32 @@ describe('Policies Page', () => {
...
+    await waitForText('Manage columns')
+    const checkbox = screen.getByTestId('checkbox-remediation')
+    fireEvent.click(checkbox)
+    screen.getByRole('button', { name: /save/i }).click()
+    await waitForText('Filter')
+    await waitForText('enforce (overridden)', true)
+    screen.getByRole('button', { name: 'Options menu' }).click()
+    await waitForText('Enforce')
+    const enforceDiv = screen.getByText('Enforce').closest('div')
+    within(enforceDiv!).getByText('1')
```
Comment (mprahl) [inline frontend/src/routes/Governance/policies/Policies.test.tsx:153]: What is this part doing?

Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/policies/Policies.test.tsx:153]: Check our `Enforce 1`

File: frontend/src/routes/Governance/policies/policy-details/PolicyDetailsOverview.test.tsx (hunk 17 above)
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/policies/policy-details/PolicyDetailsOverview.test.tsx:88]: This only test the PolicyDetailsOverview component not the whole screen `Remediation` only shows on this component... also other tests also checking like this

File: frontend/src/routes/Governance/policies/Policies.test.tsx (hunk 19 above)
Comment (mprahl) [reply frontend/src/routes/Governance/policies/Policies.test.tsx:153]: Oh yeah, I see.

File: frontend/src/routes/Governance/policies/policy-details/PolicyDetailsOverview.test.tsx (hunk 17 above)
Comment (mprahl) [reply frontend/src/routes/Governance/policies/policy-details/PolicyDetailsOverview.test.tsx:88]: I suppose it's okay since the mock data is very controlled.

**Comment by mprahl** [comment]
@yiraeChristineKim the existing functionality looks good, however, we need a way to indicate which clusters are enforced, which was my intention of the Results table having a "Remediation" column, however, it is per policy-template instead of per policy.
image

**Comment by yiraeChristineKim** [comment]
> "Remediation" column

Remediation column in result table?

File: frontend/src/routes/Governance/policies/policy-details/PolicyDetailsResults.tsx [hunk 20]
```diff
@@ -72,6 +75,7 @@ export default function PolicyDetailsResults(props: { policy: Policy }) {
             timestamp: detail?.history && detail?.history[0]?.lastTimestamp,
             policyName,
             policyNamespace,
+            remediationAction: getObjectTempRemediation(policyResponse, template),
```
Comment (yiraeChristineKim) [inline frontend/src/routes/Governance/policies/policy-details/PolicyDetailsResults.tsx:78]: I added columns

File: frontend/src/routes/Governance/common/util.tsx [hunk 21]
```diff
@@ -463,28 +464,69 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
+    return rootRA
+  }
+
+  return REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDDEN
+}
+
+export function getObjectTempRemediation(propagatedPolicy: Policy, template: PolicyTemplate | undefined): string {
+  const propagatedRA = propagatedPolicy.spec.remediationAction
+  const objectRemediation = template?.objectDefinition.spec.remediationAction
+  return propagatedRA || objectRemediation || REMEDIATION_ACTION.INFORM
```
Comment (yiraeChristineKim) [inline frontend/src/routes/Governance/common/util.tsx:529]: The definition indicates that "The policy Results tab's table should have a `Remediation` column. Add an indicator that it is overridden."  but there's isn't `overridden`  Is this right?

Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:529]: Yeah, what I meant by that was to show the actual remediation action and the user can see if it's enforce when the policy is inform, that it was overridden. We don't need to display that it's overridden.

Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:529]: It doesn't actually default to `inform`. It's possible that the user forgot to add the remediationAction field on the policy template and the policy which will eventually cause an error.  In this case, leaving it blank is correct.

File: frontend/src/routes/Governance/policies/policy-details/PolicyDetailsResults.tsx [hunk 22]
```diff
@@ -61,6 +63,7 @@ export default function PolicyDetailsResults(props: { policy: Policy }) {
           const template = templates.find(
             (template: any) => template?.objectDefinition?.metadata?.name === detail?.templateMeta?.name
           )
+          console.log(policyResponse, template)
```
Comment (mprahl) [inline frontend/src/routes/Governance/policies/policy-details/PolicyDetailsResults.tsx:66]:
```suggestion
```

File: frontend/src/routes/Governance/common/util.tsx [hunk 23]
```diff
@@ -463,28 +464,69 @@ export function getPolicySetPolicies(policies: Policy[], policySet: PolicySet) {
...
+
+  // No overrides are present
+  if (!remediationOverride) {
+    return rootRA
+  }
+
+  return REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDDEN
+}
+
+export function getObjectTempRemediation(propagatedPolicy: Policy, template: PolicyTemplate | undefined): string {
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:526]:
```suggestion
export function getPolicyTempRemediation(propagatedPolicy: Policy, template: PolicyTemplate | undefined): string {
```

File: frontend/src/routes/Governance/common/util.tsx (hunk 21 above)
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:529]: Thanks! great!

File: frontend/src/routes/Governance/policies/policy-details/PolicyDetailsResults.tsx (hunk 22 above)
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/policies/policy-details/PolicyDetailsResults.tsx:66]: Thanks for catch

**Comment by mprahl** [comment]
@yiraeChristineKim this looks good to me! I'll approve it once the CI failure is addressed.

File: frontend/src/routes/Governance/common/util.tsx [hunk 24]
```diff
@@ -492,19 +494,61 @@ export function getPolicyRemediation(policy: Policy | undefined) {
...
+
+  return rootRA.includes('informOnly')
+    ? REMEDIATION_ACTION.INFORM_INFORMONLY_ENFORCE_OVERRIDDEN
+    : REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDDEN
+}
+
+export function getPolicyTempRemediation(propagatedPolicy: Policy, template: PolicyTemplate | undefined): string {
+  const propagatedRA = propagatedPolicy.spec.remediationAction
+  const objectRemediation = template?.objectDefinition.spec.remediationAction
+  return propagatedRA || objectRemediation || ''
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:551]: This will need to change to account for `informOnly`.

File: frontend/src/routes/Governance/common/util.tsx [hunk 25]
```diff
@@ -492,19 +494,61 @@ export function getPolicyRemediation(policy: Policy | undefined) {
...
-      return 'informOnly'
+
+  if (rootRA) {
+    if (remediationAggregation.includes(REMEDIATION_ACTION.INFORM_ONLY)) {
+      rootRA = remediationSet.size > 1 ? rootRA.concat('/informOnly') : REMEDIATION_ACTION.INFORM_ONLY
+    }
+  } else {
+    rootRA = remediationAggregation
+    // empty root remediation and template remedation all empty
+    if (!rootRA) {
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:509]: It'd technically work if you had no remediationAction set on the root policy and policy templates but used an override on all clusters.

File: frontend/src/routes/Governance/common/util.tsx (hunk 24 above)
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:551]: you mean adding comment?

File: frontend/src/routes/Governance/common/util.tsx [hunk 26]
```diff
@@ -492,19 +494,61 @@ export function getPolicyRemediation(policy: Policy | undefined) {
...
+
+    if (!remediationOverride) {
+      remediationOverride = REMEDIATION_ACTION.ENFORCE
+    }
+  })
+
+  if (allOverridden) {
+    return rootRA.includes('informOnly')
+      ? REMEDIATION_ACTION.INFORMONLY_ENFORCE_OVERRIDDEN
+      : REMEDIATION_ACTION.ENFORCE_OVERRIDDEN
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:535]: What about the case where you just have a single policy template with `informOnly`? Then the override wouldn't have an effect.

File: frontend/src/routes/Governance/common/util.tsx [hunk 27]
```diff
@@ -492,19 +494,61 @@ export function getPolicyRemediation(policy: Policy | undefined) {
...
-  return remediationAggregation
+
+  // No overrides are present
+  if (!remediationOverride) {
+    return rootRA
+  }
+
+  return rootRA.includes('informOnly')
+    ? REMEDIATION_ACTION.INFORM_INFORMONLY_ENFORCE_OVERRIDDEN
+    : REMEDIATION_ACTION.INFORM_ENFORCE_OVERRIDDEN
```
Comment (mprahl) [inline frontend/src/routes/Governance/common/util.tsx:545]: Same comment as above.

File: frontend/src/routes/Governance/common/util.tsx (hunk 26 above)
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:535]: but user added enforce on root?

File: frontend/src/routes/Governance/common/util.tsx (hunk 24 above)
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:551]: It was my mistake. This is right.

File: frontend/src/routes/Governance/common/util.tsx (hunk 25 above)
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:509]: Yes

Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:509]: Can you give me the case you concern?

File: frontend/src/routes/Governance/common/util.tsx (hunk 26 above)
Comment (mprahl) [reply frontend/src/routes/Governance/common/util.tsx:535]: Right. If I have a policy with no remediation action, a single policy template with `informOnly`, and overrides set on all policies, then the enforce has no effect. It's still just `informOnly`. So maybe earlier in this function, you can just return `informOnly` and not check for overrides if all the policy templates are `informOnly`.

Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:535]: how about this line 503
```
  if (rootRA) {
    if (remediationAggregation.includes(REMEDIATION_ACTION.INFORM_ONLY)) {
      if (remediationSet.size > 1 ) {
 rootRA= rootRA.concat('/informOnly') 
       }else{
            return  REMEDIATION_ACTION.INFORM_ONLY
        }
    }
```

If only one informOnly or all informOnly, override doesn't affect at all so

File: frontend/src/routes/Governance/common/util.tsx (hunk 25 above)
Comment (yiraeChristineKim) [reply frontend/src/routes/Governance/common/util.tsx:509]: root remediation null and 0 overridden on propagated fall into here

**Review by mprahl — APPROVED** [review]
